Usecase/usecase_transaction_test.go

```diff
@@ -1,157 +1,129 @@
 package Usecase
 
 import (
+	"awesomeProject/Repo"
 	"awesomeProject/entities"
 	"errors"
+	"github.com/google/uuid"
 	"github.com/stretchr/testify/assert"
 	"testing"
 )
 
-type mockTransactionRepo struct {
-	SaveCreateTransactionFunc       func(transaction *entities.Transaction) error
-	SaveGetAllTransactionFunc       func() ([]*entities.Transaction, error)
-	GetTransactionToCreateOrderFunc func(order *entities.Order) (*entities.Transaction, error)
-}
-
-func (m *mockTransactionRepo) SaveCreateTransaction(transaction *entities.Transaction) error {
-	return m.SaveCreateTransactionFunc(transaction)
-}
-func (m *mockTransactionRepo) SaveGetAllTransaction() ([]*entities.Transaction, error) {
-	return m.SaveGetAllTransactionFunc()
-}
-func (m *mockTransactionRepo) GetTransactionToCreateOrder(order *entities.Order) (*entities.Transaction, error) {
-	return m.GetTransactionToCreateOrderFunc(order)
-}
-
 func TestTransactionUseCase_CreateTransaction(t *testing.T) {
+	mockProductRepo := new(Repo.ProductRepoMock)
+	mockTransactionRepo := new(Repo.TransactionRepoMock)
+	service := NewTransactionUseCase(mockTransactionRepo, mockProductRepo)
+	transaction := &entities.Transaction{
+		TransactionId: uuid.New(),
+		OrderAddress:  "th",
+		Items: []entities.Item{
+			{ProductId: 1, Quantity: 5, Price: 100},
+			{ProductId: 2, Quantity: 3, Price: 200},
+		},
+		TotalPrice: 1100,
+	}
+	transactionInvalidCountry := &entities.Transaction{
+		TransactionId: uuid.New(),
+		OrderAddress:  "xx",
+		Items: []entities.Item{
+			{ProductId: 1, Quantity: 5, Price: 100},
+			{ProductId: 2, Quantity: 3, Price: 200},
+		},
+		TotalPrice: 1100,
+	}
+	transactionDuplicateIdProduct := &entities.Transaction{
+		TransactionId: uuid.New(),
+		OrderAddress:  "th",
+		Items: []entities.Item{
+			{ProductId: 1, Quantity: 5, Price: 100},
+			{ProductId: 1, Quantity: 3, Price: 200},
+		},
+		TotalPrice: 1100,
+	}
 	t.Run("success", func(t *testing.T) {
-		TRepo := &mockTransactionRepo{
-			SaveCreateTransactionFunc: func(transaction *entities.Transaction) error {
-				return nil
-			},
-		}
-		PRepo := &mockProductRepo{
-			GetPriceProductsFunc: func(transaction *entities.Transaction) (*entities.Transaction, error) {
-				return &entities.Transaction{}, nil
-			},
-		}
-		service := NewTransactionUseCase(TRepo, PRepo)
-
-		err := service.CreateTransaction(
-			&entities.Transaction{
-				OrderAddress: "th",
-				Items:        []entities.Item{{ProductId: 1, Quantity: 5}, {ProductId: 2, Quantity: 3}},
-			},
-		)
+		mockProductRepo.On("GetPriceProducts", transaction).Return(transaction, nil).Once()
+		mockTransactionRepo.On("SaveCreateTransaction", transaction).Return(transaction, nil).Once()
+		result, err := service.CreateTransaction(transaction)
 		assert.NoError(t, err)
+		assert.NotNil(t, result)
+		assert.Equal(t, transaction, result)
+		mockTransactionRepo.AssertExpectations(t)
```

Comment:
You not have mockProductRepo.AssertExpectations(t) ?

Reply:
what do you mean?